### PR TITLE
feat: automatically migrate database folder, ensuring eternal backwards compatibility

### PR DESCRIPTION
**Pull request by LesnyRumcajs**
## Summary of changes

Changes introduced in this pull request:

- Introduced a database migration framework: this should ensure we no longer have to manually re-create Forest databases going forward.
- More post-migration checks would be great to implement - I can either do this here or in a follow-up PR,
- the integration test should ensure we catch breaking changes early and we always have a migration at hand (and that the person that breaks it has the golden opportunity to implement the migration!)
- implemented a PoC migration from `0.12.1`. This migration won't be performed outside of tests (unless someone manually changes the `paritydb` directory to `0.12.1`)
- adhered (mostly) to the [agreed design](https://github.com/ChainSafe/forest/blob/main/documentation/src/developer_documentation/database_migrations.md)
- Introduced a DB mode environmental variable, more explained in the Rust documentation and README.

## Reference issue to close (if applicable)

Closes https://github.com/ChainSafe/forest/issues/3203

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: documentation/src/environment_variables.md [hunk 1]
```diff
@@ -4,7 +4,23 @@ Besides CLI options and the configuration values in the configuration file,
...
+| FOREST_CAR_LOADER_FILE_IO  | 1 or true                        | false   | Load CAR files with `RandomAccessFile` instead of `Mmap` |
+| FOREST_DB_DEV_MODE         | [see here](#-forest_db_dev_mode) | current | The database to use in development mode                  |
+
+### FOREST_DB_DEV_MODE
+
+By default, Forest will create a database of its current version or try to
+migrate to it. This can be overridden with the `FOREST_DB_DEV_MODE`
+environmental variable.
+
+| Value                          | Description                                                                                                                                      |
```
Comment (jdjaustin) [inline documentation/src/environment_variables.md:19]: For some reason, this section does not display correctly in the rich diff preview.

Comment (LesnyRumcajs) [reply documentation/src/environment_variables.md:19]: It should be fine now.

**Review by jdjaustin — APPROVED** [review]

**Comment by hanabi1224** [comment]
A question about the scope of the DB migration, does it cover only parity-db or also the Forest CAR DB? See https://github.com/ChainSafe/forest/pull/3365#discussion_r1299225452

> Do note that the ForestCAR.zst format is not fixed, and re-encoding may be required.

**Comment by lemmih** [comment]
> A question about the scope of the DB migration, does it cover only parity-db or also the Forest CAR DB? See [#3365 (comment)](https://github.com/ChainSafe/forest/pull/3365#discussion_r1299225452)
> 
> > Do note that the ForestCAR.zst format is not fixed, and re-encoding may be required.
> 
> @lemmih correct me if I'm wrong, as I understand, we need to migrate forest CAR files when a breaking change has been made to ForestCAR.zst format

CAR files are as much part of the database as ParityDB is.

**Comment by LesnyRumcajs** [comment]
> A question about the scope of the DB migration, does it cover only parity-db or also the Forest CAR DB? See [#3365 (comment)](https://github.com/ChainSafe/forest/pull/3365#discussion_r1299225452)
> 
> > Do note that the ForestCAR.zst format is not fixed, and re-encoding may be required.
> 
> @lemmih correct me if I'm wrong, as I understand, we need to migrate forest CAR files when a breaking change has been made to ForestCAR.zst format

The DB migration should encompass all migrations required to have Forest transition from one version to another. At the moment we care only about the ParityDb (as it's the only one that breaks on every other update). That said, in the future we should also track things like keystore (probably should be moved to the DB? Not 100% sure, we'd need to assess security implications of this).

**Pull request merged**